cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,499 @@
+    return (int)slresult;
+}
+
+constexpr uint64_t bit_spread( const uint64_t mask, const int data )
+{
+    uint64_t rmask = binary_reverse( mask );
+    int dat = data;
+    uint64_t result = 0;
+    unsigned char lz{0};
+    while ( rmask )
+    {
+        lz = count_leading_zeros( rmask ) + 1;
+        result = result << lz | ( dat & 1 );
+        dat >>= 1, rmask <<= lz;
+    }
+    result = binary_reverse( result ) >> count_leading_zeros( mask );
+    return result;
+}
+//---------------------------------------------------------------------------//
+
+// Tile Hash Key (hash) <=> TileID
+template <typename Key, typename HashType>
+struct TileID2HashKey;
+
+template <typename Key, typename HashType>
+struct HashKey2TileID;
+
+// functions to compute hash key
+// can be rewriten in a recursive way
+template <typename Key>
+struct TileID2HashKey<Key, HashTypes::Naive>
+{
+    TileID2HashKey( std::array<int, 3> &&tnum )
+        : _tilenum( tnum )
+    {
+    }
+    TileID2HashKey( int i, int j, int k )
+        : _tilenum( i, j, k )
+    {
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( std::array<int, 3> tileid ) -> Key
+    {
+        return tileid[0] * _tilenum[1] * _tilenum[2] + tileid[1] * _tilenum[2] +
+               tileid[2];
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( int tileid_x, int tileid_y, int tileid_z ) -> Key
+    {
+        return tileid_x * _tilenum[1] * _tilenum[2] + tileid_y * _tilenum[2] +
+               tileid_z;
+    }
+
+  private:
+    std::array<int, 3> _tilenum;
+};
+
+template <typename Key>
+struct TileID2HashKey<Key, HashTypes::Morton>
+{
+    TileID2HashKey( std::array<int, 3> && ) {} // ugly
+    TileID2HashKey( int, int, int ) {}
+
+    enum : uint64_t
+    { // hand-coded now, can be improved by iterative func
+        page_zmask = ( 0x9249249249249249UL ),
+        page_ymask = ( 0x2492492492492492UL ),
+        page_xmask = ( 0x4924924924924924UL )
+    };
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( std::array<int, 3> tileid ) -> Key
+    {
+        return bit_spread( page_zmask, tileid[2] ) |
+               bit_spread( page_ymask, tileid[1] ) |
+               bit_spread( page_xmask, tileid[0] );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( int tileid_x, int tileid_y, int tileid_z ) -> Key
+    {
+        return bit_spread( page_zmask, tileid_z ) |
+               bit_spread( page_ymask, tileid_y ) |
+               bit_spread( page_xmask, tileid_x );
+    }
+};
+
+// can be rewriten in a recursive way
+template <typename Key>
+struct HashKey2TileID<Key, HashTypes::Naive>
+{
+    HashKey2TileID( std::array<int, 3> &&tnum )
+        : _tilenum( tnum )
+    {
+    }
+    HashKey2TileID( int i, int j, int k )
+        : _tilenum( i, j, k )
+    {
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, std::array<int, 3> &tileid )
```

Comment:
also eliminate the array functions in these `HashKey2TileID` specializations and use ijk instead of xyz